prestashopv1/plugin/modules/indodana/tools/IndodanaTools.php

```diff
@@ -0,0 +1,166 @@
+            $tax += ($product['price_without_reduction'] - self::getPriceWithoutReductionWithoutTax($product))
+                * $product['quantity'];
+            $deficit += self::getDeficit($product);
+        }
+
+        return round($tax) + $deficit;
+    }
+
+    public static function getProducts($cart)
+    {
+        $items = [];
+        $products = $cart->getProducts();
+        foreach ($products as $product) {
+            $items[] = [
+                'id' => $product['reference'],
+                'name' => $product['name'],
+                'price' => round(self::getPriceWithoutReductionWithoutTax($product)),
+                'quantity' => $product['quantity'],
+                'category'  => IndodanaCommon\IndodanaConstant::DEFAULT_ITEM_CATEGORY,
+            ];
+        }
+
+        return $items;
+    }
+
+    public static function getCustomerDetails($orderId)
+    {
+        $orderDetails = new Order((int) $orderId);
+        $customer = new Customer((int) $orderDetails->id_customer);
+        $invoiceDetails = new Address((int) $orderDetails->id_address_invoice);
+
+        return [
+            'firstName' => $customer->firstname,
+            'lastName' => $customer->lastname,
+            'email' => $customer->email,
+            'phone' => $invoiceDetails->phone
+        ];
+    }
+
+    public static function getBillingAddress($orderId)
+    {
+        $orderDetails = new Order((int) $orderId);
+        $invoiceDetails = new Address((int) $orderDetails->id_address_invoice);
+
+        return [
+            'firstName' => $invoiceDetails->firstname,
+            'lastName' => $invoiceDetails->lastname,
+            'address' => $invoiceDetails->address1 . ' ' . $invoiceDetails->address2,
+            'city' => $invoiceDetails->city,
+            'postalCode' => $invoiceDetails->postcode,
+            'phone' => $invoiceDetails->phone,
+            'countryCode' => Configuration::get('INDODANA_STORE_COUNTRY_CODE')
+        ];
+    }
+
+    public static function getDeliveryAddress($orderId)
+    {
+        $orderDetails = new Order((int) $orderId);
+        $deliveryDetails = new Address((int)($orderDetails->id_address_delivery));
+
+        return [
+            'firstName' => $deliveryDetails->firstname,
+            'lastName' => $deliveryDetails->lastname,
+            'address' => $deliveryDetails->address1 . ' ' . $deliveryDetails->address2,
+            'city' => $deliveryDetails->city,
+            'postalCode' => $deliveryDetails->postcode,
+            'phone' => $deliveryDetails->phone,
+            'countryCode' => Configuration::get('INDODANA_STORE_COUNTRY_CODE')
+        ];
+    }
+
+    public static function getTotalProduct($product)
+    {
+        return round(self::getPriceWithoutReductionWithoutTax($product) * $product['quantity']);
+    }
+
+    /**
+     * sometimes when rounding up/down numbers there will be deficit
+     * so that the total price calculation does not match
```

Comment:
Can you please give example for this case (maybe add in the docs)

Reply:
Sample case:
`checkout` request body:
```
{
  "merchantOrderId": "17-FZSVHEMWQ",
  "totalAmount": 3009000,
  "discountAmount": 0,
  "shippingAmount": 9000,
  "taxAmount": 272727,
  "products": [
    {
      "id": "demo_7",
      "name": "Printed Chiffon Dress",
      "price": 454545,
      "quantity": 6,
      "category": "others"
    }
  ]
  ...
}
```

totalAmount (from Prestashop) = 3009000

totalProductPrice = 454545 * 6 = 2727270
totalAmount (manually) = 2727270 + 272727 + 9000 = 3008997

because we've rounded the product price, so in the results of manual calculation there is a deficit 3

and if we send this request, we will get repsonse:
`Request Exception:"[5e96b5b4-e123-11ea-9c01-00163e014000.17-FZSVHEMWQ] Purchase transaction amount (3009000) is different from calculated amount at list of item (3008997)."`